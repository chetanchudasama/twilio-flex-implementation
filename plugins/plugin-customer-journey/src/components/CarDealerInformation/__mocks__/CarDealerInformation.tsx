import React from "react";
import { Props } from "../CarDealerInformation";

const CarDealerInformation: React.FC<Props> = (props) => <div />;

export default CarDealerInformation;
