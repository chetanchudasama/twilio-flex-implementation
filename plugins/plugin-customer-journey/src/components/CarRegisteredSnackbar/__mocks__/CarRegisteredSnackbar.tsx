import React from "react";
import { CarRegisteredSnackbarProps } from "../CarRegisteredSnackbar.Props";

export const CarRegisteredSnackbar: React.FC<CarRegisteredSnackbarProps> = (
  props
) => {
  return <div className="car-registered-snackbar">CarRegisteredSnackbar</div>;
};
