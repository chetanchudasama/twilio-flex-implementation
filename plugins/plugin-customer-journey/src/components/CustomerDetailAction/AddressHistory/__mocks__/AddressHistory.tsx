import React from "react";
import { AddressHistoryProps } from "../AddressHistory";

const AddressHistory: React.FC<AddressHistoryProps> = (props) => (
  <div className="address-history" />
);

export default AddressHistory;
