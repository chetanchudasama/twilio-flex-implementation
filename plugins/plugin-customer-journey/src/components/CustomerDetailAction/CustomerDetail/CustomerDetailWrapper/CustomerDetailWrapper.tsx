/* eslint-disable no-param-reassign */
import React, { useState } from "react";
import { Manager } from "@twilio/flex-ui";
import { CustomerDetailModel } from "@common/components";
import { preparePatchData } from "../../../../helpers/commonFunctions";
import {
  showMessage,
  showErrorMessage,
  CustomNotificationType,
} from "../../../../Notifications/index";
import { DispatchToProps, StateToProps } from "./CustomerDetailWrapper.Props";
import Loading from "../../../LoadingComponent/Loading.Container";
import CustomerProfile from "../CustomerProfile/CustomerProfile.Container";
import { AppState } from "../../../../states";
import { useApplicationService } from "../../../../services/application.service";
import { CustomerDetailWrapperStyles } from "./CustomerDetailWrapper.Styles";
import { UpdateDetailRequestModel } from "../../../../models/UpdateDetailRequestModel";

interface CustomerDetailProps {
  backToWizardStepper: () => void;
}

export type ComponentProps = StateToProps &
  DispatchToProps &
  CustomerDetailProps;

export const CustomerDetailWrapper: React.FC<ComponentProps> = ({
  applicationId,
  setCustomerDetails,
  backToWizardStepper,
}) => {
  const state: AppState = Manager.getInstance().store.getState();
  const applicationService = useApplicationService(
    state.flex.session.ssoTokenPayload.token
  );
  const [errors, setErrors] = useState<Record<string, string[]>>({});
  const [isLoading, setIsLoading] = useState(false);
  const propertyList = [
    "title",
    "firstName",
    "middleName",
    "lastName",
    "dateOfBirth",
    "maritalStatus",
    "drivingLicenceType",
    "mobileNumber",
    "workNumber",
    "homeNumber",
    "emailAddress",
  ];

  const emailRegex = /\S+@\S+\.\S+/;

  const updateProfile = async (customerProfileModel: CustomerDetailModel) => {
    let hasError = false;
    if (!customerProfileModel.firstName) {
      errors.firstName = ["The first name field is required."];
      hasError = true;
    }
    if (!customerProfileModel.lastName) {
      errors.lastName = ["The last name field is required."];
      hasError = true;
    }
    if (customerProfileModel.maritalStatus?.maritalStatusId === -1) {
      errors.maritalStatus = ["The marital status field is required."];
      hasError = true;
    }
    if (customerProfileModel.drivingLicenceType?.drivingLicenceTypeId === -1) {
      errors.drivingLicenceType = [
        "The driving license type field is required.",
      ];
      hasError = true;
    }
    if (!customerProfileModel.emailAddress) {
      errors.emailAddress = ["The email field is required."];
      hasError = true;
    }
    if (
      customerProfileModel.emailAddress &&
      !emailRegex.test(customerProfileModel.emailAddress)
    ) {
      errors.emailAddress = ["The email field must be a valid email."];
      hasError = true;
    }
    if (!customerProfileModel.dateOfBirth) {
      hasError = true;
    }
    setErrors(errors);

    if (!hasError) {
      // filter properties to be updated
      const customerDetail = Object.keys(customerProfileModel)
        .filter((key) => propertyList.includes(key))
        .reduce((object, key) => {
          return {
            ...object,
            [key]: (customerProfileModel as any)[key],
          };
        }, {});

      try {
        setIsLoading(true);

        // prepare patch request data
        const updateRequestList: UpdateDetailRequestModel[] = [];
        Object.keys(customerDetail).forEach((key: string, index: number) => {
          updateRequestList.push(
            preparePatchData(key, Object.values(customerDetail)[index])
          );
        });

        await applicationService
          .updateBaseApplicationDetail(applicationId, updateRequestList)
          .then(() => {
            // show success notification
            showMessage(
              CustomNotificationType.SuccessNotification,
              "Customer detail updated successfully!"
            );
            // update customer detail in redux
            Object.keys(customerDetail).forEach((key, index) =>
              setCustomerDetails(
                key as any,
                Object.values(customerDetail)[index]
              )
            );
            // wait till detail updated in redux
            setTimeout(() => {
              // return to wizard
              backToWizardStepper();
            }, 0);
          });
      } catch (error) {
        setErrors({});
        showErrorMessage(
          "Error in updating customer detail, please try again!",
          "",
          true
        );
      } finally {
        setIsLoading(false);
      }
    }
  };

  return (
    <CustomerDetailWrapperStyles>
      {isLoading && <Loading />}
      <CustomerProfile errorList={errors} updateProfile={updateProfile} />
    </CustomerDetailWrapperStyles>
  );
};
