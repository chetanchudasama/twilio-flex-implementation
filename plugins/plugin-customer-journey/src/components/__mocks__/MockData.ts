import { WorkerAttributes } from "@twilio/flex-ui";
import { namespace } from "../../states";

const workerAttributes: WorkerAttributes = {
  contact_uri: "client:test",
  email: "test@test.com",
  full_name: "test user",
  image_url: "img_url",
  roles: ["admin"],
  agentId: 1,
};

const conferenceSource = {
  source: {
    participants: [
      {
        callSid: "callSid",
        participantType: "customer",
        workerSid: "",
      },
      {
        callSid: "callSid",
        participantType: "worker",
        workerSid: "workerSid",
      },
    ],
  },
};

const conferenceState = new Map();
conferenceState.set("taskSid", conferenceSource);

export const flexManagerMock = jest.fn().mockReturnValue({
  workerClient: {
    source: {},
    sid: "workerSid",
    name: "test user",
    fullName: "test user",
    activityName: "Offline",
    attributes: workerAttributes,
    dateUpdated: new Date(),
    isAvailable: false,
    activityDuration: "",
    dateStatusChanged: new Date(1).toISOString(),
    activity: {
      available: false,
      name: "Offline",
      sid: "OfflineSid",
    },
  },
  store: {
    getState: jest.fn().mockReturnValue({
      flex: {
        worker: {
          attributes: workerAttributes,
        },
        session: {
          ssoTokenPayload: {
            token: "ssoToken",
          },
        },
        conferences: {
          states: conferenceState,
        },
      },
      [namespace]: {
        phoneNumberList: {
          phoneNumberList: [
            {
              friendlyName: "Sales",
              phoneNumber: "+1234567891",
              queueSid: "QUEUE_SID",
            },
          ],
        },
      },
    }),
  },
  strings: {
    DIALPADExternalTransferSearchInputLabel: "Search By Name or PhoneNumber",
    DIALPADExternalTransferHoverOver: "Add",
    DIALPADExternalTransferPopupHeader: "External Transfer",
  },
  serviceConfiguration: {
    outbound_call_flows: {
      default: {
        caller_id: "CALLER_ID",
      },
    },
  },
});
