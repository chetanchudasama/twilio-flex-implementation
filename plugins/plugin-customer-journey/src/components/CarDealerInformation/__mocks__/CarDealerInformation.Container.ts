import CarDealerInformation from "./CarDealerInformation";

export default CarDealerInformation;
