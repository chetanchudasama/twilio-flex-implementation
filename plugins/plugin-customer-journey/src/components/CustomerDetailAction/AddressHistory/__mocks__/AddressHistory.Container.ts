import AddressHistory from "./AddressHistory";

export default AddressHistory;
