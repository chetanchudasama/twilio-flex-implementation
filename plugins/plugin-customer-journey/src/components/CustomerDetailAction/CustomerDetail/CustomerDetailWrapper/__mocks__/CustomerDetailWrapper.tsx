import React from "react";
import { ComponentProps } from "../CustomerDetailWrapper";

const CustomerDetailWrapper: React.FC<ComponentProps> = (props) => (
  <div className="customer-detail-wrapper" />
);

export default CustomerDetailWrapper;
