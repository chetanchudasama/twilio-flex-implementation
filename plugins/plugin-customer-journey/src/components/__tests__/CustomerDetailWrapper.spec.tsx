import React from "react";
import { create } from "react-test-renderer";
import flushPromises from "flush-promises";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { Manager } from "@twilio/flex-ui";
import { showErrorMessage, showMessage } from "../../Notifications";
import { CustomerDetailWrapper } from "../CustomerDetailAction/CustomerDetail/CustomerDetailWrapper/CustomerDetailWrapper";
import { flexManagerMock } from "../__mocks__/MockData";

jest.mock("../../services/application.service");

jest.mock("@common/components", () => {
  const components = jest.requireActual("@common/components");
  return {
    ...components,
    CustomerDetailModel: jest.fn().mockImplementation(() => {
      return { customerDetailModel: () => {} };
    }),
  };
});

jest.mock("../../Notifications", () => {
  return {
    showErrorMessage: jest.fn(() => {}),
    showMessage: jest.fn(() => {}),
    CustomNotificationType: { SuccessNotification: "successNotification" },
  };
});

const mockUpdateBaseApplicationDetail = jest.fn();

jest.mock("../../services/application.service", () => ({
  useApplicationService: jest.fn(() => ({
    updateBaseApplicationDetail: mockUpdateBaseApplicationDetail,
  })),
}));

jest.mock(
  "../CustomerDetailAction/CustomerDetail/CustomerProfile/CustomerProfile"
);
jest.mock(
  "../CustomerDetailAction/CustomerDetail/CustomerProfile/CustomerProfile.Container"
);

const defaultProps = {
  applicationId: 1,
  setCustomerDetails: jest.fn(() => {}),
  backToWizardStepper: jest.fn(() => {}),
};

const props = defaultProps;

Manager.getInstance = flexManagerMock;

jest.mock("../LoadingComponent/Loading.Container", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>Loading</div>;
    },
  };
});

describe("CustomerDetailWrapper", () => {
  jest.useFakeTimers();
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it("renders correctly", async () => {
    const wrapper = create(<CustomerDetailWrapper {...props} />).toJSON();
    await flushPromises();

    expect(wrapper).toMatchSnapshot();
  });

  it("should update customer detail and show success notification", async () => {
    mockUpdateBaseApplicationDetail.mockImplementationOnce(() =>
      Promise.resolve(undefined)
    );

    await act(async () => {
      const wrapper = mount(<CustomerDetailWrapper {...props} />);
      wrapper.update();
      await flushPromises();
      wrapper.update();

      const button = wrapper.find(".update-detail-btn").at(0);
      button.simulate("click");
      await flushPromises();
      wrapper.update();

      expect(mockUpdateBaseApplicationDetail).toHaveBeenCalledTimes(1);

      expect(showMessage).toHaveBeenCalledWith(
        "successNotification",
        "Customer detail updated successfully!"
      );

      expect(props.setCustomerDetails).toHaveBeenCalled();

      expect(setTimeout).toHaveBeenCalled();
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 0);
    });
  });

  it("should show error message, if error while update customer details", async () => {
    mockUpdateBaseApplicationDetail.mockImplementationOnce(() =>
      Promise.reject()
    );

    await act(async () => {
      const wrapper = mount(<CustomerDetailWrapper {...props} />);
      wrapper.update();
      await flushPromises();
      wrapper.update();

      const button = wrapper.find(".update-detail-btn");
      button.simulate("click");
      await flushPromises();
      wrapper.update();

      expect(showErrorMessage).toHaveBeenCalledWith(
        "Error in updating customer detail, please try again!",
        "",
        true
      );
    });
  });
});
