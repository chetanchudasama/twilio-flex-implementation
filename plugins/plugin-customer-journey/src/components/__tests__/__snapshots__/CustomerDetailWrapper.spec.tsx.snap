// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetailWrapper renders correctly 1`] = `
<div
  className="css-0"
>
  <div
    className="customer-profile"
  >
    <button
      className="update-detail-btn"
      onClick={[Function]}
      type="button"
    >
      Update Details
    </button>
  </div>
</div>
`;
